### PR TITLE
Support offset and size for read

### PR DESCRIPTION
**Pull request by zhaohuxing**

### CODE REVIEW

File: storage.go [hunk 1]
```diff
@@ -473,6 +473,7 @@ func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairSt
 	input := &s3.GetObjectInput{
 		Bucket: aws.String(s.name),
 		Key:    aws.String(rp),
+		Range:  aws.String(fmt.Sprintf("bytes=%d-%d", opt.Offset, opt.Offset+opt.Size)),
```
Comment (Xuanwo) [inline storage.go:476]: `Range` is `[a, b]`, maybe we need to use `opt.Offset+opt.Size-1` for end?

Comment (Xuanwo) [reply storage.go:476]: BTW, we need to check offset and size before we use them.

- What if the offset is missing?
- What if the size is missing?

Comment (zhaohuxing) [reply storage.go:476]: > `Range` is `[a, b]`, maybe we need to use `opt.Offset+opt.Size-1` for end?
Got

Comment (zhaohuxing) [reply storage.go:476]: > BTW, we need to check offset and size before we use them.
> 
> * What if the offset is missing?
> * What if the size is missing?

Both offset and size are not pointer types, so we have to check whether size is greater than 0

File: storage.go [hunk 2]
```diff
@@ -474,6 +474,9 @@ func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairSt
 		Bucket: aws.String(s.name),
 		Key:    aws.String(rp),
 	}
+	if opt.Size > 0 {
```
Comment (Xuanwo) [inline storage.go:477]: Please use `opt.HasXxxx` to check `size` and `offset`.

It's OK to eliminate the check for `offset`, but please fill comments here for the reason.

Comment (zhaohuxing) [reply storage.go]: Modified, Please Review。

File: storage.go [hunk 3]
```diff
@@ -474,6 +474,9 @@ func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairSt
 		Bucket: aws.String(s.name),
 		Key:    aws.String(rp),
 	}
+	if opt.HasOffset && opt.HasSize {
```
Comment (Xuanwo) [inline storage.go:477]: What if the user only inputs `size` or `offset`? (NOTE: they are both valid.)

- size=x, offset=0 -> read from 0 to x
- size=0, offset=x -> read from x to the end
- size=x, offset=y -> read from x to x+y

We need to handle all of them.

Comment (zhaohuxing) [reply storage.go:477]: OK

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**